File: Cpp/AdventOfCode/src/2015/d17_EggnogStorage.cpp
#pragma once

#include "2015/d17_EggnogStorage.h"

SOLUTION(2015, 17) {
    constexpr u32 CountCombinations(const std::vector<u32>&containers, size_t index, u32 remaining) {
        auto next = containers[index];

        if (remaining < next) {
            return 0;
        }

        remaining -= next;
        if (remaining == 0) {
            return 1;
        }

        u32 result = 0;
        for (auto i = index + 1; i < containers.size(); i++) {
            result += CountCombinations(containers, i, remaining);
        }

        return result;
    }

    constexpr u32 CountCombinations(std::vector<u32>&containerSizes, u32 initial) {
        containerSizes.push_back(0); // :(
        std::sort(containerSizes.begin(), containerSizes.end());
        return CountCombinations(containerSizes, 0, initial);
    }

    constexpr u32 GetMinSize(const std::vector<u32>&containers, size_t index, u32 remaining, u32 currentLength) {
        auto next = containers[index];
        if (remaining < next) {
            return 0;
        }
        remaining -= next;
        if (remaining == 0) {
            return currentLength + 1;
        }

        u32 best = 999;
        for (auto i = index + 1; i < containers.size(); i++) {
            u32 length = GetMinSize(containers, i, remaining, currentLength + 1);
            if (length > 0) {
                best = std::min(best, length);
            }
        }

        return best;
    }

    constexpr u32 CountCombinationsWithMaxContainers(const std::vector<u32>&containers, size_t index, u32 remaining, u32 currentLength, const u32 maxSize) {
        if (currentLength >= maxSize) {
            return 0;
        }

        auto next = containers[index];
        if (remaining < next) {
            return 0;
        }
        remaining -= next;
        if (remaining == 0) {
            return 1;
        }

        u32 result = 0;
        for (auto i = index + 1; i < containers.size(); i++) {
            result += CountCombinationsWithMaxContainers(containers, i, remaining, currentLength + 1, maxSize);
        }

        return result;
    }

    constexpr u32 CountSmallContainers(std::vector<u32>&containers, u32 initial) {
        containers.push_back(0); // :(
        std::sort(containers.begin(), containers.end());

        u32 size = GetMinSize(containers, 0, initial, 0);

        return CountCombinationsWithMaxContainers(containers, 0, initial, 0, size);
    }

    PART_ONE() {
        auto containers = ParseLinesAsNumbers<u32>(Lines);
        return Constexpr::ToString(CountCombinations(containers, 150));
    }
    PART_TWO() {
        auto containers = ParseLinesAsNumbers<u32>(Lines);
        return Constexpr::ToString(CountSmallContainers(containers, 150));
    }

    TESTS() {
        std::vector<u32> containers = {20, 5, 10, 5, 15};
        std::sort(containers.begin(), containers.end());

        if (CountCombinations(containers, 25) != 4) return false;
        if(GetMinSize(containers, 0, 25, 0) != 3) return false; //0 10 15
        if(CountSmallContainers(containers, 25) != 3) return false;

        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d17_eggnog_storage.rs
pub fn part_one(lines: &Vec<String>) -> String {
    let containers: Vec<i32> = lines.iter().map(|l| l.parse::<i32>().unwrap()).collect();
    count_combinations_initial(containers, 150).to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    let containers: Vec<i32> = lines.iter().map(|l| l.parse::<i32>().unwrap()).collect();
    count_small_containers(containers, 150).to_string()
}

fn count_combinations(containers: &Vec<i32>, index: usize, remaining: i32) -> i32 {
    match remaining - containers[index] {
        r if r < 0 => 0,
        r if r == 0 => 1,
        r => (index + 1..containers.len()).map(|i| count_combinations(containers, i, r)).sum()
    }
}

fn count_combinations_initial(mut containers: Vec<i32>, initial: i32) -> i32 {
    containers.push(0);
    containers.sort();
    count_combinations(&containers, 0, initial)
}

fn get_min_size(containers: &Vec<i32>, index: usize, remaining: i32, current_length: i32) -> i32 {
    match remaining - containers[index] {
        r if r < 0 => 0,
        r if r == 0 => current_length + 1,
        r => (index + 1..containers.len()).map(|i| get_min_size(containers, i, r, current_length + 1)).filter(|v| *v > 0).min().unwrap_or(0)
    }
}

fn count_combos_with_max_containers(containers: &Vec<i32>, index: usize, remaining: i32, current_length: i32, max_size: i32) -> i32 {
    if current_length >= max_size {
        return 0;
    }
    match remaining - containers[index] {
        r if r < 0 => 0,
        r if r == 0 => 1,
        r => (index + 1..containers.len()).map(|i| count_combos_with_max_containers(containers, i, r, current_length + 1, max_size)).sum()
    }
}

fn count_small_containers(mut containers: Vec<i32>, initial: i32) -> i32 {
    containers.push(0);
    containers.sort();
    let min_size = get_min_size(&containers, 0, initial, 0);
    count_combos_with_max_containers(&containers, 0, initial, 0, min_size)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn count_combinations_test() {
        let mut containers = vec![20, 5, 10, 5, 15];
        containers.sort();

        assert_eq!(4, count_combinations_initial(containers, 25));
    }

    #[test]
    fn get_min_size_test() {
        let mut containers = vec![20, 5, 10, 5, 15];
        containers.sort();

        assert_eq!(2, get_min_size(&containers, 0, 25, 0));
    }

    #[test]
    fn count_small_containers_tests() {
        let mut containers = vec![20, 5, 10, 5, 15];
        containers.sort();

        assert_eq!(3, count_small_containers(containers, 25));
    }
}